### ISSUE TITLE
Run(git ls-remote -q origin refs/heads/master): context deadline exceeded

### ISSUE THREAD

**Issue by mrc-rius** [kubernetes/git-sync#498]
Hi,

I have my git-sync working in production for some days already and yesterday we experienced a break in the container suddenly.

The error of the log is:

`E0224 07:25:52.769877      12 main.go:506] "msg"="too many failures, aborting" "error"="Run(git ls-remote -q origin refs/heads/master): context deadline exceeded: { stdout: "", stderr: "" }" "failCount"=0`

This breaks the container and after that all the cluster fells down as its an essential container.

I didn't find specifically this error but I assume this operation is internal of the container.

If I setup `GIT_SYNC_MAX_SYNC_FAILURES=-1` I guess it won't break but I didn't tried if the git pull will still works as it should.

What could be the best approach?
Thank you

**Comment by thockin** [comment]
It's telling you that it could not talk to the upstream server in the given
timeout.  Either the server is dead or something is slow on network.  If it
has been working, it sounds like a transient error and simply restarting
should eventually recover.

What does "This breaks the container" mean?  Does it not recover after the
upstream comes back?

**Comment by mrc-rius** [comment]
Thanks @thockin 

Could be interesting to increase the time out using `GIT_SYNC_TIMEOUT`?
Yesterday afternoon it repeated.

I have the container mark as essential, because it is the provdider of the code. When this fail the docker contaner crashes and dies. After that the AWS ECS tasks dies and a new one is raised. It creates a 3 minutes downtime.

**Comment by thockin** [comment]
I'm not super familiar with ECS semantics, but if you are regularly timing
out, I would indeed set the timeout higher and/or set the failcount higher

**Comment by mrc-rius** [comment]
Okey, great, thanks. Does this affect the git pull statement?

**Comment by thockin** [comment]
No, this just limits how long we are willing to wait before giving up.  If
things move quickly, we will not waste time.

Is this a particularly large repo, by chance?  Maybe with a lot of commit
history?

You might also try --depth to limit hosuch we sync.

**Comment by mrc-rius** [comment]
I don't know when to classify a repo as a big one. It has 14k commits.

--depth specifies the number of commits that are pulled or what does it mean exactly? I guess that will reduce the volume of data downloaded, but there is some risk of bad sync?

**Comment by thockin** [comment]
The --timeout flag covers the whole process - including the clone/fetch AND the garbage-collection.  I'm looking at ways to make this more flexible, but the short answer is "increase the timeout" or "decrease the depth" if you are regularly hitting timeouts.

**Comment by mrc-rius** [comment]
Okei, thanks! I will setup `GIT_SYNC_DEPTH=1000` as we always sync repo with the latests commit. So it should be more than enough. And also setup `GIT_SYNC_TIMEOUT=1200`
